pkgs/applications/graphics/imv/default.nix

```diff
@@ -1,61 +1,110 @@
-{ asciidoc
+{ stdenv
+, lib
+, fetchFromGitHub
+, asciidoc
 , cmocka
 , docbook_xsl
-, fetchFromGitHub
+, libxslt
 , fontconfig
-, freeimage
+, meson
+, ninja
+, pkgconfig
 , icu
+, pango
+, inih
+, withWindowSystem ? "all"
+, xorg
+, libxkbcommon
 , libGLU
-, libheif
+, wayland
+, withBackends ? [ "freeimage" "libtiff" "libjpeg" "libpng" "librsvg" "libnsgif" "libheif" ]
+, freeimage
+, libtiff
 , libjpeg_turbo
 , libpng
 , librsvg
-, libtiff
-, libxkbcommon
-, libxslt
 , netsurf
-, pango
-, pkgconfig
-, stdenv
-, wayland
+, libheif
 }:
 
+let
+  windowSystems = {
+    x11 = {
+      buildInputs = [ libGLU xorg.libxcb xorg.libX11 ];
+      # libxcbcommon-x11 is contained in libxcbcommon
+    };
+    wayland = {
+      buildInputs = [ wayland ];
+    };
+    all = {
+      buildInputs = windowSystems.x11.buildInputs
+        ++ windowSystems.wayland.buildInputs;
+    };
+  };
+
+  backends = {
+    inherit freeimage libtiff libpng librsvg libheif;
+    libjpeg = libjpeg_turbo;
+    inherit (netsurf) libnsgif;
+  };
+
+  backendFlags = builtins.map
+    (b: if builtins.elem b withBackends
+        then "-D${b}=enabled"
+        else "-D${b}=disabled")
+    (builtins.attrNames backends);
+in
+
+# check that given window system is valid
+assert lib.assertOneOf "withWindowSystem" withWindowSystem
+  (builtins.attrNames windowSystems);
+# check that every given backend is valid
+assert builtins.foldl' lib.and true (builtins.map
```

Comment:
You can simplify this using the `all` function:
```
builtins.all 
  (b: lib.assertOneOf "each backed" b withBackends)
  (builtins.attrNames backends)
```

Reply:
Thanks!